cpp/src/traversal/mg/frontier_expand_kernels.cuh

```diff
@@ -0,0 +1,292 @@
+        thread_frontier[thread_frontier_count++] = destination;
+      }
+    }
+    bool is_last_iter = (i == (num_iter - 1));
+    bool is_nth_iter  = (i % EdgesPerThread == 0);
+    // Write to frontier every EdgesPerThread iterations
+    // Or if it is the last iteration of the for loop
+    if (is_nth_iter || is_last_iter) {
+      write_to_frontier(thread_frontier,
+                        thread_frontier_count,
+                        block_frontier,
+                        &block_frontier_count,
+                        output_frontier,
+                        &block_write_offset,
+                        output_frontier_count);
+      thread_frontier_count = 0;
+    }
+    edge_offset += blockDim.x;
+  }
+}
+
+template <int BlockSize, int EdgesPerThread, typename vertex_t, typename edge_t, typename Operator>
+__global__ void kernel_per_vertex(edge_t *offsets,
+                                  vertex_t *indices,
+                                  vertex_t *input_frontier,
+                                  vertex_t input_frontier_count,
+                                  vertex_t vertex_begin,
+                                  vertex_t *output_frontier,
+                                  edge_t *output_frontier_count,
+                                  Operator op)
+{
+  vertex_t current_vertex_index = 0;
+  __shared__ edge_t block_write_offset;
+  __shared__ vertex_t block_frontier[BlockSize * EdgesPerThread];
+  __shared__ int block_frontier_count;
+
+  edge_t stride = blockDim.x * gridDim.x;
+  vertex_t thread_frontier[EdgesPerThread];
+
+  while (current_vertex_index < input_frontier_count) {
+    vertex_t source           = input_frontier[current_vertex_index];
+    edge_t beg_block_offset   = offsets[source] + (blockIdx.x * blockDim.x);
+    edge_t end_block_offset   = offsets[source + 1];
+    int i                     = 0;
+    int thread_frontier_count = 0;
+    for (edge_t block_offset = beg_block_offset; block_offset < end_block_offset;
+         block_offset += stride) {
+      if (block_offset + threadIdx.x < end_block_offset) {
+        vertex_t destination = indices[block_offset + threadIdx.x];
+        if (op(source + vertex_begin, destination)) {
+          thread_frontier[thread_frontier_count++] = destination;
+        }
+      }
+      bool is_last_iter = (block_offset + blockDim.x >= end_block_offset);
+      bool is_nth_iter  = (i % EdgesPerThread == 0);
+      if (is_nth_iter || is_last_iter) {
+        write_to_frontier(thread_frontier,
+                          thread_frontier_count,
+                          block_frontier,
+                          &block_frontier_count,
+                          output_frontier,
+                          &block_write_offset,
+                          output_frontier_count);
+        thread_frontier_count = 0;
+      }
+      ++i;
+    }
+    ++current_vertex_index;
+  }
+}
+
+template <typename VT, typename ET, typename WT, typename Operator>
```

Comment:
Please consider lowercase and more descriptive typenames `<name>_t.`

Reply:
Change `Operator` to `operator_t` as well?

Reply:
Sure, why not.